### ISSUE TITLE
关于NormHead的疑问

### ISSUE THREAD

**Issue by piekey1994** [baichuan-inc/Baichuan2#19]
```
class NormHead(nn.Module):
    def __init__(self, hidden_size, vocab_size, bias=False):
        super().__init__()
        self.weight = nn.Parameter(torch.empty((vocab_size, hidden_size)))
        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        self.first_flag = True

    def forward(self, hidden_states):
        if self.training:
            norm_weight = nn.functional.normalize(self.weight)
        elif self.first_flag:
            self.first_flag = False
            self.weight = nn.Parameter(nn.functional.normalize(self.weight))
            norm_weight = self.weight
        else:
            norm_weight = self.weight
        return nn.functional.linear(hidden_states, norm_weight)
```
1.当eval的时候，weight会被重新赋值，但是如果我eval后继续训练，是不是会有问题
2.在eval的过程中新创建这个weight是不是只是为了加速，其实train和eval都用norm_weight = nn.functional.normalize(self.weight)是不是也可以，不会有什么影响

**Comment by daxia-daxia** [comment]
You are right. creating new parameter in eval mode is just a speedup trick! You can use norm_weight = nn.functional.normalize(self.weight) for all mode, no harm except a little bit slowness.

**Comment by FrankWhh** [comment]
请问为啥重新赋值能加速？

**Comment by daxia-daxia** [comment]
> 请问为啥重新赋值能加速？

Normalization is omitted in this case.

**Comment by nrailgun** [comment]
测试过 NormHead 一旦代替 linear，HBM 就飙上去了，原地 OOM。

**Comment by jameswu2014** [comment]
> 有相关资料吗？我理解推理就是前项计算，融合算子可以加速，重新赋值为啥能加速

只有第一次forward会计算，后面就直接赋值了，不需要再对weight做normalize了

**Comment by jeffchy** [comment]
发现多次eval loss的时候损失很大，是否是因为当eval的时候，weight会被重新赋值，但是如果我在训练过程中连续eval，是不是会有问题？

**Comment by David-Lee-1990** [comment]
> 发现多次eval loss的时候损失很大，是否是因为当eval的时候，weight会被重新赋值，但是如果我在训练过程中连续eval，是不是会有问题？

训练过程中多次eval，并不会每次都重新赋值吧，只是第一次eval的时候。
